### ISSUE TITLE
pg-cdc: Duplicates in source if storaged is killed and restarted

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#13749]
### What version of Materialize are you using?

69c7269

### How did you install Materialize?

Docker image

### What is the issue?

If storaged is killed and restarted during a pg-cdc test, the data in the source will start showing both up-to-date records and old records (that came from a snapshot?).

To reproduce:
1. clone `https://github.com/philip-stoev/materialize.git`, branch `checks-pg-sources`
2. run
```
cd test/platform-checks
./mzcompose run  default --scenario KillStoraged
```

This will run a bunch of testdrive fragments while killing and restarting storageds at several points during the test. It will then fail as follows:

```
> SELECT f1, f2, SUM(LENGTH(f3)) FROM t1A GROUP BY f1, f2;
--
  | rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 112ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s 6s 10s 15s 22s 33s 49s 74s 75s
  | <stdin>:2:1: error: non-matching rows: expected:
  | [["A", "8", "102400"], ["B", "8", "102400"], ["C", "7", "102400"], ["D", "6", "102400"], ["E", "5", "102400"], ["F", "4", "102400"], ["G", "3", "102400"], ["H", "2", "102400"]]
  | got:
  | [["A", "3", "102400"], ["A", "8", "204800"], ["B", "3", "102400"], ["B", "8", "204800"], ["C", "2", "102400"], ["C", "7", "204800"], ["D", "6", "204800"], ["E", "5", "204800"], ["F", "4", "204800"], ["G", "3", "204800"], ["H", "2", "204800"]]
```

Or in other words, the expected data looks like this:

```
postgres=#  SELECT f1, f2, SUM(LENGTH(f3)) FROM t1 GROUP BY f1, f2;
 f1 | f2 |  sum   
----+----+--------
 A  |  8 | 102400
 D  |  6 | 102400
 E  |  5 | 102400
 H  |  2 | 102400
 B  |  8 | 102400
 G  |  3 | 102400
 C  |  7 | 102400
 F  |  4 | 102400
```
Or one row per letter, and sum = 102400 always

But the actual data looks like this:

```
materialize=>  SELECT f1, f2, SUM(LENGTH(f3)) FROM t1A GROUP BY f1, f2;
 f1 | f2 |  sum   
----+----+--------
 A  |  3 | 102400
 A  |  8 | 204800
 B  |  3 | 102400
 B  |  8 | 204800
 C  |  2 | 102400
 C  |  7 | 204800
 D  |  6 | 204800
 E  |  5 | 204800
 F  |  4 | 204800
 G  |  3 | 204800
 H  |  2 | 204800
```
Or more than one row per letter, and sum = 204800. I think the double sum indicates that some rows were ingested twice. The fact that there are two rows per letter indicates that the same record appears in multiple versions.

If a view was created earlier in the test, like `t1A` , before multiple storaged restarts, it contains both anomalies. If the view was created later, and only subjected to one storaged restart, like `t1D`, it will contain only the sum=204800 anomaly and not the other.

The issue does not seem repeatable if only the pg-cdc-related queries from the framework are run. 

### Relevant log output

```source
No interesting messages apart from those about lost connectivity and rehydration.
```

**Comment by philip-stoev** [comment]
I am adding the C-Triage tag to this bug in order for it to enter back into the triage system. There has been zero movement (tags, priority, assigment, anything) on this ticket for 14 days.

**Comment by JLDLaughlin** [comment]
cc @uce

**Comment by aljoscha** [comment]
Sorry for the slow response! @petrosagg Could you please take a look at this once you're back and have time. You should be the person that is most familiar with a) Postgres and b) the restart logic.

**Comment by petrosagg** [comment]
Writing up some context of why this issue happens.

Every time a source ingestion is instantiated we pick a frontier to restart at which we then pass all the way down to the source implementation. The first time a pg source is instantiated that translates to the start LSN being zero. Now, there is no way for us to actually read updates from LSN zero, that is long gone, so the source implementation treats that as a signal that its output is empty and so it needs to perform a snapshot at some LSN **that is picked at that moment**.

Now suppose that the source picked LSN 1000 and produced a snapshot at that LSN which we then reclocked at time `12:00pm`. Now ff that ingestion pod crashes and a new one is scheduled we must re-calculate the resumption frontier to give it. When this issue was created the storage controller would re-issue the stored commands to the newly spun up storaged instance and those commands would say `resume_upper = {0}`, just like the first time.

The source implementation, just like the first time, treated this as a signal to generate a snapshot. However, because some time has passed the LSN that the second ingestion **picked in that moment** is now 2000. So the ingestion generates a snapshot at LSN 2000, reclocks it to say `12:01pm` and writes it to the output shard. Now we've ended up with an output shard that has duplicate rows.

In current `main` this is much harder to happen because we have changed the rehydration logic to recalculate the resume upper every time a new `storaged` is started. However, we're still not immune from this and BIR won't fix it either. It's possible that we schedule a `storaged` pod at resume upper 0, have that pod go in a network partition, and then start an additional storage pod because we think the original one is dead.

Now we have two concurrent storage pods, running the same postgres ingestion, both initialized at resume upper 0. Sometimes this will produce duplicate rows.

The crux of the issue is that we assume that upstream sources never compact under our feet unless we give them permission to do so but postgres violates this by not being able to provide the same snapshot at the same LSN a second time.

One way to fix this issue would be to differentiate between source ingestions that just resume from source ingestions that start over with a snapshot in time. In the latter case we can render the self-correcting sink dataflow fragment at the end to ensure that the end state is identical to the desired state. Concretely I think `SourceReader` needs to gain an API that given a resumption frontier it tells us whether it can actually resume operation from it or if must re-snapshot the data because the upstream compacted (or that it can't do either because the upstream data is gone).

**Comment by uce** [comment]
@petrosagg Thank you for the write-up. Would the `SourceReader` changes you have in mind conflict with the ongoing work in BIR or Postgres demuxing? I'm trying to think through whether this can be tackled concurrently or should wait on the BIR or Postgres demuxing.

**Comment by guswynn** [comment]
@uce I don't think the API that @petrosagg suggests conflicts in any fundamental way with BIR, though it may require some light rebasing on top of each other if they are done at the same time. Not sure about pg demuxing

**Comment by petrosagg** [comment]
@uce it could be done concurrently but I foresee high chance of merge conflicts since there is a lot of activity in the ingestion pipeline currently.

Today I thought of a quick fix that will get us correctness until we can implement a better solution. When we render the pg ingestion pipeline make it so that we use `compare_and_append` when writing to the output shard (instead of plain `append`) and crash if we ever detect a race. Doing this will prevent a concurrent snapshot from being inserted because in the failure case the sinking part of the ingestion dataflow sees a batch spanning `[0..1000]` in one case and `[0..2000]` in the other case where the snapshot happens at 1000 and 2000 respectively and plain `append` happily merges the two into one stream.

By crashing we'll force the controller to reschedule a pod if needed which will go through the rehydration logic which will calculate a fresh resume_frontier which will be at the upper of the output shard.

**Comment by uce** [comment]
Thanks for your answers, folks!

Two take-aways:
* Let's do the one-line fix Petros outlined (if there are no technical concerns)
* Let's not start anything else before BIR and PG demux are done. We can revisit whether we need a better fix after those are done.

WDYT?

**Comment by guswynn** [comment]
@petrosagg don't we already use `compare_and_append`? https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/render/persist_sink.rs#L180

**Comment by philip-stoev** [comment]
I rebased and the test no longer fails, so I  merged it to "main". I am leaving this ticket open in case any further discussion and/or fixes needs to happen.

**Comment by aljoscha** [comment]
> @petrosagg don't we already use `compare_and_append`? https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/render/persist_sink.rs#L180

Yes! I think this is fixed for now.

@petrosagg do you foresee any problems if we'd switch to the self-correcting `persist_sink`? Two instances of `storaged` might fight about what should be in the shard but that would settle down once one of them is stopped. wdyt?

**Comment by guswynn** [comment]
@aljoscha I believe we need to decide on which of these properties for "multiple processes writing to same shard" we want right?
- We always chose a winner after some bounded time
- We only choose a winner if the other processes shut off

**Comment by aljoscha** [comment]
yes! And I'm afraid we can't yet put an answer in place, but renditions might give us the lever we need because it can "fence out" an older instance of the code that for some reason has kept running.

**Comment by petrosagg** [comment]
> do you foresee any problems if we'd switch to the self-correcting persist_sink?

We'd need to have a way to differentiate between "source is providing a snapshot" from "source is resuming" which we don't do currently. But I'm starting to become increasingly skeptical of the self-correcting sink, at least for ingestion. The alternative design which is to schedule a dataflow with the assumption that it's the only one baked in and have it fail when that's not true sounds much more appealing. In that case the recovery is dead-simple, just re-schedule it and a new resumption frontier will be calculated.

Right now the above would also imply a crash of `storaged` but this can be improved in the future in a way that we gracefully catch the error, tear down the dataflow and restart it within the same process.

**Comment by aljoscha** [comment]
> But I'm starting to become increasingly skeptical of the self-correcting sink, at least for ingestion.

Me too! The PoC-PR I have for that doesn't work, and won't work without subtle tweaking. We might have to do a simple thing that still writes things in parallel because it seems we're already hitting throughput problems in sources.

**Comment by aljoscha** [comment]
@petrosagg Just to wrap up this one: because we're currently using `compare_and_append`, the bug shouldn't occur anymore, as you suggested above and as Philip confirmed empirically. Right?

**Comment by petrosagg** [comment]
That's right! Closing

**Comment by aljoscha** [comment]
:heart_on_fire:
